src/Transients.cc

```diff
@@ -287,36 +296,36 @@ void
 Transients::abandon(const StoreEntry &e)
 {
     assert(e.mem_obj && map);
-    map->freeEntry(e.mem_obj->xitTable.index); // just marks the locked entry
-    CollapsedForwarding::Broadcast(e);
+    // avoid useless broadcasts
+    bool stateChanged = false;
+    map->freeEntry(e.mem_obj->xitTable.index, &stateChanged); // just marks the locked entry
+    if (stateChanged)
+        CollapsedForwarding::Broadcast(e, true);
     // We do not unlock the entry now because the problem is most likely with
     // the server resource rather than a specific cache writer, so we want to
     // prevent other readers from collapsing requests for that resource.
 }
 
-bool
-Transients::abandoned(const StoreEntry &e) const
-{
-    assert(e.mem_obj);
-    return abandonedAt(e.mem_obj->xitTable.index);
-}
-
-/// whether an in-transit entry at the index is now abandoned by its writer
-bool
-Transients::abandonedAt(const sfileno index) const
+void
+Transients::status(const StoreEntry &entry, bool &aborted, bool &waitingToBeFreed) const
 {
     assert(map);
-    return map->readableEntry(index).waitingToBeFreed;
+    assert(entry.mem_obj);
+    const auto idx = entry.mem_obj->xitTable.index;
+    const auto &anchor = collapsedWriter(entry) ?
+        map->writeableEntry(idx) : map->readableEntry(idx);
+    aborted = EBIT_TEST(anchor.basics.flags, ENTRY_ABORTED);
```

Comment:
> Any better ideas?

Having two duplicated flags looks not very well. Moreover, what if in future we would need to read one more 'flag' from anchor.basics.flags in another reader? Again add such 'atomic' duplicate?

I would like to suggest another approach:

```
aborted = false;
// Since a shared entry gets ENTRY_ABORTED only within StoreMap::abortWriting(),
// (where anchor.writing() becomes false), we should not check this value before.
// Also note that trying to read ENTRY_ABORTED for being written shared entry
// may cause an unpredictable result, because anchor.basics.flags is not 'atomic'.
if (!anchor.writing() && EBIT_TEST(anchor.basics.flags, ENTRY_ABORTED))
      aborted = true;
waitingToBeFreed = anchor.waitingToBeFreed;
```

Reply:
> Having two duplicated flags looks not very well

I think we can avoid any duplication of flags in my proposal: The new atomic flag will be used without any relationship or reference to the unused(?) same-name bit in basics.flags. And we can name the new atomic flag `writerGone` or something like that to minimize the implication that the two are flags are duplicated.

Moreover, there is actually some extra danger in using ENTRY_ABORTED like we do in the current PR code. I suspect that some higher level code may assert if it gets an ENTRY_ABORTED entry from the cache. Our code probably copies shared `basic.flags` into local StoreEntry objects blindly, which (in rare cases) might result in a cache reader getting a ENTRY_ABORTED StoreEntry hit and asserting. The code should handle later hit aborts correctly (because they can always happen on incomplete entries), but finding an already aborted hit in the cache, especially for a "complete" entry, may be problematic. This concern alone may be a sufficient reason to _not_ use ENTRY_ABORTED for our purposes, even if the atomicity issue does not exist!

Finally, we should consider removing `basics.flags` as effectively unused, but probably not in this PR because that decision is not trivial.

```
// Since a shared entry gets ENTRY_ABORTED only within StoreMap::abortWriting(),
// (where anchor.writing() becomes false), we should not check this value before.
if (!anchor.writing() && EBIT_TEST(anchor.basics.flags, ENTRY_ABORTED))
      aborted = true;
```

This trick misses the (rare) case where the flag has been set (before the check) but the write lock has not been released (until after the check). This problem is not critical because the affected reading worker should receive another notification later, when the writer finishes updating its state. At that time, the reader will detect the aborted entry correctly. Thus, we can get away with this by adding an XXX.

The question is: Why should we use tricks and add XXXs when we can use an atomic flag instead? I do not think we should if the earlier discussion addresses your "duplicated flags" concern.

Reply:
Done at 8cd83030. Does it meet your expectation?

> Can we place that second aborted flag so that sizeof(StoreMapAnchor) does not change on 32bit and 64bit platforms?

I placed it just after waitingToBeFreed. I assume that since sizeof(waitintToBeFreed) == sizeof (writerHalted) == 1, these fields are not aligned. The following data field (uint64_t key[2]) should be aligned with 3-byte offset before changes and with 2-byte offset after changes. The sizeof(StoreMapAnchor) should be the same in both cases, I think. In my environment it is 96 bytes.